engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsANode.java

```diff
@@ -0,0 +1,35 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "is_a",
+    description = "Checks type of a node.",
+    autoRegister = false)
+public abstract class IsANode extends Node {
+
+  public abstract boolean execute(@AcceptsError Object value, Object type);
+
+  public static IsANode build() {
+    return IsANodeGen.create();
+  }
+
+  @Specialization
+  boolean doTypeCheck(Object value, Type type, @CachedLibrary(limit = "10") TypesLibrary types) {
+    var t = types.getType(value);
+    return t == type;
+  }
+
+  @Fallback
+  boolean doAny(Object value, Object type) {
+    return value == type;
```

Comment:
This is very surprising check, but needed. Otherwise for example [this check](https://github.com/enso-org/enso/blob/0ad70c63328c4e165f4b837bdf94c8c6177662de/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso#L140) yields wrong result!

Reply:
I'm wondering though if this is the expected semantics of `is_a`? I.e. should we say `is_a` _without_ this check is wrong or maybe _this usage_ is wrong?

My intuition is that `is_a` checks if a value is of a given type. The linked example however checks for value, not type.

Reply:
IMHO it will be easier for the users when `x.is_a Typ` will be aligned with `case x of _ : Typ ->` - there will be one concept to grasp. Otherwise we will have two very similar concepts that will differ in a subtle way - IMO the worst situation because the user can falsely think they do the same and then get shot in the foot when they actually differ.

Reply:
There is, I think, only one instance where old `is_a` was shining (although I'm not sure it is currently used within the stdlib, but I think the feature itself is useful) - being able to check if a given value is an instance of a specific atom _constructor_ without depending on this constructor's arity.

For example:
```
type Foo
    Bar a b c
    Baz x y

hmm v =
    v.is_a Foo.Bar
```

This way we can check if the value is of the specific `Foo.Bar` variant. This cannot be done with the `_ : Typ` match as it will match _all_ `Foo` instances:
```
hmm v = case v of
   _ : Foo -> "I'm either a Bar or also Baz"
   _ -> "I'm sth else"
```

And with regular pattern match, we will need to know the arity:
```
hmm v = case v of
    Foo.Bar _ _ _ -> "I'm Foo.Bar!"
   _ -> "I'm sth else, maybe Foo.Baz, maybe not Foo at all :)"
```

Sometimes the arity changes and without good refactoring tools it is a pain to check all pattern matches (well the compiler shouts about it, so it's not the end of the world). But sometimes arity is just an implementation detail and shouldn't be exposed in places like this, just for the sake of cleanliness/elegance.

Now - it seems that this new implementation is losing this ability altogether (with the exception of 0-ary constructors which you link in the above example).

IMO if we want to keep the 0-ary, we should keep it for all kinds of constructors.

And I'd vote to have this kind of functionality as I (subjectively) think it can be sometimes useful (although it is not indispensable).

However, I'm really not sure if `is_a` is it's right place - I still think that ensuring `is_a` is aligned with `case-of` on types is the way to go. Maybe we can split this out into a separate functionality? Possibly on `Meta.Constructor` or something similar?

Anyway, if we are not using it currently, it may not be worth spending time on implementing it - but maybe worth at least creating a Pivotal issue to track that we _may_ want to have it - or just remember at the back of our head that such possibility can potentially be implemented - as it can sometimes make our life easier.

Regardless, I'd keep that separate from `is_a`.

Reply:
> when x.is_a Typ will be aligned with case x of _ : Typ -> - there will be one concept to grasp

That's the case now. Covered by `MetaIsATest.consistencyWithCase` test.

Reply:
> Anyway, if we are not using it currently, it may not be worth spending time on implementing it - but maybe worth at least creating a Pivotal issue to track that we may want to have it - or just remember at the back of our head that such possibility can potentially be implemented - as it can sometimes make our life easier.

Reincarnated as #5792